### PR TITLE
Squash wip improvement

### PR DESCRIPTION
**Pull request by gregorriegler**
Adds new `mob done --squash-wip` option, which does a `mob done` but also runs `squashWip` beforehand.
As a result, when you use `mob done --squash-wip`, you can keep the manual commits you made on the wip branch while getting rid of all the wip commits.

`squashWip` had to be tweaked for this to fixup subsequent wip commits that had no following manual commit. And it also undoes a final wip commit, so that the user is forced to commit those changes manually.

The environment var MOB_DONE_SQUASH is made backward compatible, while the `.mob` config file does not support activating --squash-wip yet.

Questions:

- Should we now already discard the `mob squash-wip` command after all?
- Can you find any corner cases I missed?

### CODE REVIEW

**Comment by hollesse** [comment]
It seems to me that the adjusted configuration option squash done can not be set from the configuration files. I just saw the parsing of a Boolean and not a string, like you did it for the parsing of the environment variables.

**Comment by gregorriegler** [comment]
> It seems to me that the adjusted configuration option squash done can not be set from the configuration files. I just saw the parsing of a Boolean and not a string, like you did it for the parsing of the environment variables.

Correct

File: mob.go [hunk 1]
```diff
@@ -467,6 +475,20 @@ func setBoolean(s *bool, key string, value string) {
 	debugInfo("Overwriting " + key + " =" + strconv.FormatBool(boolValue))
 }
 
+func setMobDoneSquash(configuration *Configuration, key string, value string) {
```
Comment (simonharrer) [inline mob.go:478]: Hm, not sure whether we really benefit from the enum here. Just using a string would probably be simpler.

File: squash_wip_test.go [hunk 2]
```diff
@@ -297,3 +346,31 @@ squash c51a56d manual commit
 	result, _ := ioutil.ReadFile(input)
 	equals(t, expected, string(result))
 }
+
+func wipCommit(t *testing.T, configuration Configuration, filename string) {
+	start(configuration)
+	createFile(t, filename, "irrelevant")
```
Comment (simonharrer) [inline squash_wip_test.go:352]: could you use `"contentIrrelevant"` in all cases? Little bit easier to read in github PRs.

File: squash_wip.go [hunk 3]
```diff
@@ -32,13 +24,24 @@ last commit must be a manual commit`)
...
+	sayInfo("the history of your '" + currentWipBranch.String() + "' branch has been rewritten to combine all wip commits, if possible with their following manual commits:")
 	sayEmptyLine()
 	sayLastCommitsWithMessage(currentBaseBranch.String(), currentWipBranch.String())
-	sayEmptyLine()
-	sayTodo("to finally put the changes into the base branch preserving the resulting commits, call:", configuration.mob("done --no-squash"))
+	if lastCommitIsWipCommit(configuration) { // last commit is wip commit
+		sayInfo("undoing the final wip commit and staging its changes:")
+		git("reset", "--soft", "HEAD^")
+	}
+	git("push", "--force", "--no-verify")
```
Comment (simonharrer) [inline squash_wip.go:36]: You need to get the info from the configuration, see `configuration.gitHooksOption()`

File: squash_wip.go [hunk 4]
```diff
@@ -32,13 +24,24 @@ last commit must be a manual commit`)
 		mobExecutable()+" squash-wip --git-editor",
 		mobExecutable()+" squash-wip --git-sequence-editor",
 	)
-	silentgit("rebase", "-i", "--keep-empty", mergeBase)
+	git("rebase", "-i", "--keep-empty", mergeBase)
```
Comment (simonharrer) [inline squash_wip.go:27]: having a sayInfo describing what is about to happen before the rebase might be helpful

File: squash_wip.go [hunk 5]
```diff
@@ -32,13 +24,24 @@ last commit must be a manual commit`)
...
 	setEnvGitEditor(originalGitEditor, originalGitSequenceEditor)
-	sayInfo("the history of your '" + currentWipBranch.String() + "' branch has been rewritten to combine all wip commits with their following manual commits:")
+	sayInfo("the history of your '" + currentWipBranch.String() + "' branch has been rewritten to combine all wip commits, if possible with their following manual commits:")
```
Comment (simonharrer) [inline squash_wip.go:29]: I don't like this "if possible", but I don't have a great idea at hand...

File: squash_wip.go (hunk 4 above)
Comment (simonharrer) [inline squash_wip.go:27]: Instead of "-i" we should write "--interactive" (always use the more readable parameter)

File: mob.go (hunk 1 above)
Comment (gregorriegler) [reply mob.go:478]: It is a string. It's just masked with its own type. This helps with refactoring. Also, it makes sure that we don't accidentally use invalid values, so it isolates this invariant. And on top of that, it helps me with coding, as I don't have to remember what exactly the string was, I just use auto-completion. I understand, that the translation code from env variable to this new Type is a little complicated. But it would be as complicated if it was a string. So if you asked me, I'm all for the enumish Type wrapping.

File: squash_wip_test.go (hunk 2 above)
Comment (gregorriegler) [reply squash_wip_test.go]: I agree

File: mob.go (hunk 1 above)
Comment (gregorriegler) [reply mob.go:490]: I just improved on the code that parses the DoneSquash configuration. Its odd. Even though they are just strings, I cannot compare them to strings.

File: mob.go [hunk 6]
```diff
@@ -27,29 +27,49 @@ var (
...
+const (
+	Squash    DoneSquash = "squash"
+	NoSquash             = "no-squash"
+	SquashWip            = "squash-wip"
+)
+
+func doneSquash(value string) DoneSquash {
+	switch value {
+	case "false":
+		fallthrough
```
Comment (simonharrer) [inline mob.go:41]: you can have two cases comma separated...

**Comment by simonharrer** [comment]
Can you add a changelog entry for the next version?

File: squash_wip.go [hunk 7]
```diff
@@ -32,13 +24,25 @@ last commit must be a manual commit`)
 		mobExecutable()+" squash-wip --git-editor",
 		mobExecutable()+" squash-wip --git-sequence-editor",
 	)
-	silentgit("rebase", "-i", "--keep-empty", mergeBase)
+	sayInfo("rewriting the history of the '" + currentWipBranch.String() + "' branch to squash wip commits but keep manual commits.")
```
Comment (simonharrer) [inline squash_wip.go:27]: shorter wording: `rewriting history of %s: squashing wip commits while keeping manual commits`

**Comment by simonharrer** [comment]
LGTM

**Comment by gregorriegler** [comment]
I did the final two changes from my Phone 😅.
Can't do the changelog entry right now. If you want it now, feel free to add it.

**Comment by gregorriegler** [comment]
All set

**Comment by gregorriegler** [comment]
Ok, I just added a note that the old squash-wip command is deprecated.

File: README.md [hunk 8]
```diff
@@ -174,7 +174,8 @@ Basic Commands(Options):
 
 Experimental Commands:
   squash-wip                             Combines wip commits in wip branch with subsequent manual commits to leave only manual commits.
-                                         ! Works only if all wip commits have the same wip commit message !
+                                         !Deprecated! As of version 2.6.0 this command is obsolete as the behaviour is added to 'done --squash-wip' which is more convenient.
+                                         
     [--git-editor]                       Not intended for manual use. Used as a non-interactive editor (GIT_EDITOR) for git.
     [--git-sequence-editor]              Not intended for manual use. Used as a non-interactive sequence editor (GIT_SEQUENCE_EDITOR) for git.
```
Comment (simonharrer) [inline README.md:180]: Would remove this alltogether. We don't have to explain when removing experimental commands.

**Comment by simonharrer** [comment]
We should remove the squash-wip command, or add a deprecation warning when this command is called.

**Comment by gregorriegler** [comment]
I was not able to remove the command entirely, as we use it as an editor in the squash-wip feature. But I left only the editor part and removed it from the docs.

**Comment by simonharrer** [comment]
Awesome. Thanks for this great PR! :-)
